metagraph/tests/graph/all/test_dbg_helpers.cpp

```diff
@@ -126,6 +132,62 @@ build_graph<DBGBitmap>(uint64_t k,
     return graph;
 }
 
+void writeFastaFile(const std::vector<std::string>& sequences, const std::string& outputFilename) {
+    std::ofstream fastaFile(outputFilename);
+
+    if (!fastaFile.is_open()) {
+        std::cerr << "Error: Unable to open the output file." << std::endl;
+        return;
+    }
+
+    for (size_t i = 0; i < sequences.size(); ++i) {
+        fastaFile << ">"<< "\n" << sequences[i] << "\n";
+    }
+
+    fastaFile.close();
+}
+template <>
+std::shared_ptr<DeBruijnGraph>
+build_graph<DBGSSHash>(uint64_t k,
+                       std::vector<std::string> sequences,
+                       DeBruijnGraph::Mode mode) {
+    if (sequences.empty())
+        return std::make_shared<DBGSSHash>(k, mode);
+
+    // use DBGHashString to get contigs for SSHash
+    auto string_graph = build_graph<DBGHashString>(k, sequences, mode);
+
+    std::vector<std::string> contigs;
+    size_t num_kmers = 0;
+    string_graph->call_sequences([&](const std::string &contig, const auto &path) {
+        contigs.push_back(contig);
+        num_kmers += path.size();
+    }, 1, mode != DeBruijnGraph::BASIC);
+
+    if (contigs.empty())
+        return std::make_shared<DBGSSHash>(k, mode);
+
+    std::string dump_path = "../tests/data/sshash_sequences/contigs.fa";
+    writeFastaFile(contigs, dump_path);
+
+    std::shared_ptr<DBGSSHash> graph;
+    try {
+        graph = std::make_shared<DBGSSHash>(dump_path, k, mode);
+    } catch (const std::runtime_error &e) {
+        if (strcmp(e.what(), "each partition must contain more than one key: use less partitions") == 0) {
+            return build_graph<DBGHashOrdered>(k, sequences, mode);
+        } else {
+            throw e;
+        }
+    }
+
+    if (mode == DeBruijnGraph::PRIMARY)
+        return std::make_shared<CanonicalDBG>(
```

Comment:
Why?

Reply:
A PRIMARY-mode graph only has one orientation for each k-mer, so we use CanonicalDBG as a wrapper class to gain access to both orientations of each k-mer while only storing one.

Reply:
Eh? I assumed the name `CanonicalDBG` meant it's supposed to be operated while working in `mode == CANONICAL`. Should there be some kind of assert in DBGSSHash then that PRIMARY mode is not supported directly? It looks somewhat confusing...

Reply:
You can still build a PRIMARY-mode graph and do k-mer set operations on it, so I wouldn't exactly say that it's not supported. It's just not useful for directly querying.